cpp/src/arrow/util/enum.h

```diff
@@ -0,0 +1,120 @@
+                   ? false
+                   : l[0] == '\0' ? true : CaseInsensitiveEquals(l + 1, r + 1, limit - 1);
+}
+
+constexpr bool CaseInsensitiveEquals(util::string_view l, util::string_view r) {
+  return l.size() == r.size() && CaseInsensitiveEquals(l.data(), r.data(), l.size());
+}
+
+static_assert(CaseInsensitiveEquals("a", "a"), "");
+static_assert(CaseInsensitiveEquals("Ab", "ab"), "");
+static_assert(CaseInsensitiveEquals("Ab ", "ab", 2), "");
+static_assert(CaseInsensitiveEquals(util::string_view{"Ab ", 2}, "ab"), "");
+
+constexpr const char* SkipWhitespace(const char* raw) {
+  return *raw == '\0' || !IsSpace(*raw) ? raw : SkipWhitespace(raw + 1);
+}
+
+constexpr const char* SkipNonWhitespace(const char* raw) {
+  return *raw == '\0' || IsSpace(*raw) ? raw : SkipNonWhitespace(raw + 1);
+}
+
+constexpr size_t TokenSize(const char* token_start) {
+  return SkipNonWhitespace(token_start) - token_start;
+}
+
+constexpr size_t NextTokenStart(const char* raw, size_t token_start) {
+  return SkipWhitespace(SkipNonWhitespace(raw + token_start)) - raw;
+}
+
+static_assert(CaseInsensitiveEquals(SkipWhitespace("  a"), "a"), "");
+
+using StringConstant = const char* const&;
+
+template <StringConstant Raw, size_t... Offsets>
+struct EnumTypeImpl {
+  static constexpr const char* raw = SkipWhitespace(Raw);
+
+  static constexpr int size = sizeof...(Offsets);
+
+  static constexpr util::string_view values[sizeof...(Offsets)] = {
+      {raw + Offsets, TokenSize(raw + Offsets)}...};
+
+  static constexpr int GetIndex(util::string_view repr, int i = 0) {
+    return i == size ? -1
+                     : CaseInsensitiveEquals(values[i], repr) ? i : GetIndex(repr, i + 1);
+  }
+};
+
+template <StringConstant Raw, size_t... Offsets>
+constexpr util::string_view const
+    EnumTypeImpl<Raw, Offsets...>::values[sizeof...(Offsets)];
+
+template <StringConstant Raw, bool IsEnd = false,
+          size_t MaxOffset = SkipWhitespace(Raw) - Raw, size_t... Offsets>
+struct EnumTypeBuilder
+    : EnumTypeBuilder<Raw, Raw[NextTokenStart(Raw, MaxOffset)] == '\0',
+                      NextTokenStart(Raw, MaxOffset), Offsets..., MaxOffset> {};
+
+template <StringConstant Raw, size_t MaxOffset, size_t... Offsets>
+struct EnumTypeBuilder<Raw, true, MaxOffset, Offsets...> {
+  using ImplType = EnumTypeImpl<Raw, Offsets...>;
+};
+
+template <StringConstant Raw>
+struct EnumType : EnumTypeBuilder<Raw>::ImplType {
+  constexpr EnumType() = default;
+  constexpr explicit EnumType(int i) : index{i < this->size ? i : -1} {}
+  constexpr explicit EnumType(util::string_view repr) : index{this->GetIndex(repr)} {}
+
+  constexpr bool operator==(EnumType other) const { return index == other.index; }
+  constexpr bool operator!=(EnumType other) const { return index != other.index; }
+
+  std::string ToString() const { return this->values[index].to_string(); }
+  constexpr explicit operator bool() const { return index != -1; }
```

Comment:
nit: either the constructor should also guard for negative indices or this check should be for any negative index

Reply:
I'll guard for any index outside [0, size)